### DISCUSSION TITLE
Disallowing different amount of parameters

### DISCUSSION THREAD

**Discussion by eithed** [spaze/phpstan-disallowed-calls#154] [Q&A] [answered]
I'd like to disallow different methods, with different messages, dependent upon the number of parameters used. For example, if I've got function:
```
function foo($bar = '', $xyz = ''){}
```

I'd like to display:
- if `foo()` is used anywhere, display `Don't use foo`
- if `foo(*)` is used anywhere, display `Don't use foo with parameter $bar`
- if `foo(*,*)` is used anywhere, display `Don't use foo with parameters $bar and $xyz`

Is this possible at all?

**Comment by spaze** [comment, marked as answer]
Hi, unfortunately this is not currently possible. The message is defined only on a function/method/... level and is the same for any number of parameters used.

 I may add a param-level message that would override the main level message if the call is disallowed due to a parametr being used/missing. I think you may then be able to create a config that would do what you need.

**Comment by spaze** [comment]
I have created #155 so I don't forget 😅 Thanks for the idea.

**Reply by eithed** [reply]
Just fyi for the code:
```
config();
config('foo');
```

And rules:
```
parameters:
	disallowedFunctionCalls:
		-
			function: 'config()'
			message: 'foo'
		-
			function: 'config()'
			message: 'bar'
			allowParamsAnywhereAnyValue:
				-
					position: 1
```

Only first instance is reported as invalid with `Calling config() is forbidden, bar`

For:
```
parameters:
	disallowedFunctionCalls:
		-
			function: 'config()'
			message: 'bar'
			allowParamsAnywhereAnyValue:
				-
					position: 1
		-
			function: 'config()'
			message: 'foo'
```
Both instances are reported with: `Calling config() is forbidden, foo`

---

I think one of the above configurations could be adapted to introduce the proposed - I guess the first setup with:
- first line defining rules for any parameters
- second line defining rules that doesn't have first parameter

Thus `config()` would result in `bar` message, and `config('foo');` would result in `foo` message; I must admit though wrapping your head around such setup and ensuring backward compatibility  might warrant additional parameters. Anyhow, food for thoughts :)

**Reply by spaze** [reply]
Hey @eithed, sorry it took me this long , and hopefully you'll still find useful, but I think I've now added something that should help you write the rules as you'd like, see #232. See also [the test added there](https://github.com/spaze/phpstan-disallowed-calls/pull/232/files#diff-17a5da51f32769e5a5c0dd4201988d442202d61ffd450236f429f9b15df711b4R24-R56).

I think the configuration that would work as you'd like looks like this, please note the sort of reverse order where the most specific rule with 2 params is at the top, followed by a 1-param rule, then the most general one with no params:

```neon
parameters:
	disallowedFunctionCalls:
		-
			function: 'config()'
			message: 'with two params'
			disallowParamsAnyValue:
				-
					position: 1
				-
					position: 2
		-
			function: 'config()'
			message: 'with one param'
			disallowParamsAnyValue:
				-
					position: 1
		-
			function: 'config()'
			message: 'no params'
```

If you can, and still feel like you have a use for that, could you please see if it works for your case before I merge it?

**Reply by eithed** [reply]
No worries! It's your project and I'm happy to see the idea considered :)

Am currently away from the codebase, but if it works as you've described then it does meet my needs. Thank you so much!
